### PR TITLE
Add foundational register model classes

### PR DESCRIPTION
**Pull request by tudortimi**
Add support for foundational UVM register layer classes: `uvm_reg_block`, `uvm_reg_map`, `uvm_reg` and `uvm_reg_field`.

### CODE REVIEW

**Comment by tudortimi** [comment]
I wanted to get your feedback on a couple of things, before I continue with the work.

**Comment by tudortimi** [comment]
I noticed you like to put comments with references to the standard. I added refs for both the classes and the methods. I noticed, though, that not all existing classes have the refs. What would you like to see moving forward?

**Comment by tudortimi** [comment]
I noticed that you have `raise error_classes.UVMNotImplemented(...)` for methods defined in the standard, but that aren't yet implemented. There are quite a few methods in the classes I want to add. Is it ok to add the exceptions in a subsequent PR?

**Comment by tudortimi** [comment]
Shall I add type annotations to method declarations?

**Comment by tudortimi** [comment]
I snuck an [EditorConfig](https://editorconfig.org/) file on the branch. I'm trying to use more of it. Do you have any objections? Do you want it in a separate PR?

**Comment by raysalemi** [comment]
> I noticed you like to put comments with references to the standard. I added refs for both the classes and the methods. I noticed, though, that not all existing classes have the refs. What would you like to see moving forward?

That practice went through some evolution, and I need to standardize.  I'd like to have the comment reference the spec and the docstring be useful in an IDE.  In many cases I put the standard reference in the docstring, which is useless. 

I have to admit that I don't know the first thing about the register layer. I should learn it so I can help out if you'd like a second pair of eyes.

**Comment by raysalemi** [comment]
> I snuck an [EditorConfig](https://editorconfig.org/) file on the branch. I'm trying to use more of it. Do you have any objections? Do you want it in a separate PR?

Thank you!  I did not know this existed.

**Comment by raysalemi** [comment]
> I snuck an [EditorConfig](https://editorconfig.org/) file on the branch. I'm trying to use more of it. Do you have any objections? Do you want it in a separate PR?

You can add it to this PR if you like. Either way works for me.

**Comment by tudortimi** [comment]
> > I noticed you like to put comments with references to the standard. I added refs for both the classes and the methods. I noticed, though, that not all existing classes have the refs. What would you like to see moving forward?
> 
> That practice went through some evolution, and I need to standardize. I'd like to have the comment reference the spec and the docstring be useful in an IDE. In many cases I put the standard reference in the docstring, which is useless.
> 
> I have to admit that I don't know the first thing about the register layer. I should learn it so I can help out if you'd like a second pair of eyes.

Actually wanted to ask about docstring. So, copy it from the standard?

**Comment by tudortimi** [comment]
I'd like to put the PR back into draft state, 'cause it's far from ready, but I don't know how.

**Comment by raysalemi** [comment]
> Shall I add type annotations to method declarations?

I don't want to that. 

I think one of the wonders of Python is its typelessness and I make a big deal about that in my blog posts, and in my upcoming book.  Adding type annotations would confuse the teaching message.

**Comment by tudortimi** [comment]
> > Shall I add type annotations to method declarations?
> 
> I don't want to that.
> 
> I think one of the wonders of Python is its typelessness and I make a big deal about that in my blog posts, and in my upcoming book. Adding type annotations would confuse the teaching message.

Type annotations are for code checking tools, not for the interpreter.

**Comment by raysalemi** [comment]
BTW. I find your style of many small commits to be very helpful. I tend to squash all commits into one big commit, but I typically work alone. Your approach of showing them step-by-step is excellent!

**Comment by tudortimi** [comment]
> BTW. I find your style of many small commits to be very helpful. I tend to squash all commits into one big commit, but I typically work alone. Your approach of showing them step-by-step is excellent!

We typically use "merge commit" as a merge strategy (actually rebase + merge, but we use Bitbucket). It gives the best results, IMO. Granular history, for readability + annotations, and coarse history (with 'git log --first-parent') for good overview.

**Comment by raysalemi** [comment]
@tudortimi , I presume that I wait until you change the status from "Draft" before merging this PR, right?

**Comment by tudortimi** [comment]
Yep, got a lot more to do. Drafts are for early feedback.

**Comment by tudortimi** [comment]
Another question.

There's a function, `uvm_reg::get_fields`. Its signature is `virtual function void get_fields(ref uvm_reg_field fields[$])`. How would you handle it?

IMO, this is just a lazy substitute for the fact that SV doesn't easily allow one to return arrays. The devs should have used the tried and true idiom of defining a return type using an extra `typedef`. Such a method signature, that uses pseudo-returns, makes the client code difficult to write.

I would do:
```
def get_fields():
    result = []
    // ...
    return result
```

**Comment by raysalemi** [comment]
My rule is to always refactor to use Python's power. So would do what makes sense for Python. In this case return a list.

File: pyuvm/s18_register_model.py
Comment (themperek) [inline pyuvm/s18_register_model.py:58]: Maybe one can add:
```python3
parent = property(get_parent)
````
so it is more Python-like access avalaible?

Comment (raysalemi) [reply pyuvm/s18_register_model.py:58]: I made a design decision keep the `get_*` rather than convert them to properties. 

I've never see the `property()` function, I've always used the decorator. 

What does the function do?

Comment (themperek) [reply pyuvm/s18_register_model.py:58]: You can keep `get_*` I guess make sense for compatibility.

If you add `property(get_parent)` one can do `obj.parent` also and this will call `obj.get_parent()`. 

It is up to you. Bit matter of teste.

Some more: https://www.programiz.com/python-programming/property

Comment (raysalemi) [reply pyuvm/s18_register_model.py:58]: Thanks for the link.  That's excellent in that it keeps the get_* and set_* that the UVM programmer expects but also supports the Python writer's instinct.  

If we add it we should do it across the board in one release so that the user isn't frustrated by finding that some set/get pairs don't have a property.

**Comment by tudortimi** [comment]
This PR is already quite big, so I'd rather stop at implementing the introspection related methods. Code for values can be added subsequently.

**Comment by tudortimi** [comment]
As already discussed, I would propose merging using a merge commit. (This is, of course, after any eventual changes you would like made.)

**Review by raysalemi — APPROVED** [review]
This all looks great!

**Comment by tudortimi** [comment]
Thanks for incorporating this PR. I see you used rebase, though. This puts each commit in a "linear" path in the history. Have a look at the history for SVUnit, where I use merge commits: This way, the commits in a PR stay "together" and can be treated as a logical group. If you clone the repo and do `git log --first-parent`, you'll only see a list of PR merges, which is a good overview, but it's always possible to see the entire history with granular commits.

**Comment by tudortimi** [comment]
The history on GitHub will always show all commits, unfortunately. Bitbucket is better in this respect, as it shows the lines for the merge commits as well.

**Comment by raysalemi** [comment]
> Thanks for incorporating this PR. I see you used rebase, though. This puts each commit in a "linear" path in the history. Have a look at the history for SVUnit (https://github.com/svunit/svunit/commits/master), where I use merge commits: This way, the commits in a PR stay "together" and can be treated as a logical group. If you clone the repo and do `git log --first-parent`, you'll only see a list of PR merges, which is a good overview, but it's always possible to see the entire history with granular commits.

Ah. I see that now.  Thanks for the tip!  I'll do it the other way in the future.

**Pull request merged**
